From: popcornmix <popcornmix@gmail.com>
Date: Fri, 13 Jul 2012 18:34:18 +0100
Subject: [PATCH] Fix for quitting X hang. Interrupted mailbox reads are not
 something we can reliably recover from, so down_interruptable is not a safe
 call.

---
 arch/arm/mach-bcm2708/vcio.c | 12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/arch/arm/mach-bcm2708/vcio.c b/arch/arm/mach-bcm2708/vcio.c
index 799a0ac..3874051 100644
--- a/arch/arm/mach-bcm2708/vcio.c
+++ b/arch/arm/mach-bcm2708/vcio.c
@@ -119,14 +119,10 @@ static int mbox_read(struct vc_mailbox *mbox, unsigned chan, uint32_t *data28)
 	if (mbox->magic != MBOX_MAGIC)
 		rc = -EINVAL;
 	else {
-		if (down_interruptible(&mbox->sema[chan]) == 0) {
-			*data28 = MBOX_DATA28(mbox->msg[chan]);
-			mbox->msg[chan] = 0;
-			rc = 0;
-		} else {
-			/* The wait was interrupted */
-			rc = -EINTR;
-		}
+		down(&mbox->sema[chan]);
+		*data28 = MBOX_DATA28(mbox->msg[chan]);
+		mbox->msg[chan] = 0;
+		rc = 0;
 	}
 	return rc;
 }
